# Week 1: Linear Regression - Instructor Demo Notebook

**Purpose:** Complete, polished version of live coding demo for instructor use

**Dataset:** California Housing

**Time Allocation:** 40 minutes (1:35-2:15)

---

## 📋 Instructor Notes

**Before starting:**
- Run all cells to verify they execute without errors
- Check that plots display correctly
- Review teaching tips in markdown cells
- Have Section 6 of teaching guide open for reference

**Timing guide:**
- Cell 1-2: 5 min (Imports & Load)
- Cell 3: 5 min (Split)
- Cell 4: 5 min (Train)
- Cell 5: 5 min (Predict)
- Cell 6: 7 min (Evaluate)
- Cell 7: 8 min (Residual Plot)
- Wrap-up: 5 min

**⚠️ Teaching Tips:**
- Type slowly - students are following along
- Narrate as you type
- Pause after each cell to explain output
- Check chat frequently
- Ask "Everyone got this?" every 5 minutes

---

## Cell 1: Setup & Imports (5 minutes)

### Teaching Script:
> "First thing in any ML project: bring in our tools. Let me explain each import as I type..."

### What to Emphasize:
- This is the standard pattern for every ML project
- Each import has a specific purpose
- Students should copy this pattern

In [4]:
# Week 1: Linear Regression Demo
# California Housing Dataset

# Standard imports for ML pipeline
from sklearn.datasets import fetch_california_housing
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

# For Jupyter notebooks (if plots don't show)
%matplotlib inline

print("✅ All imports successful!")

✅ All imports successful!


### Expected Output:
```
✅ All imports successful!
```

### Troubleshooting:
- If `ModuleNotFoundError`: Student needs to install scikit-learn
- Quick fix: `!pip install scikit-learn` then restart kernel

---

## Cell 2: Load Data (5 minutes)

### Teaching Script:
> "Now we load real data. This is the California Housing dataset - 20,000+ houses from 1990 census. Our goal: predict median house value."

### Key Points to Make:
- Always check data shapes immediately
- X.shape = (samples, features)
- All features are continuous (no encoding needed)

In [5]:
# Load the California Housing dataset
print("Loading California Housing dataset...")
data = fetch_california_housing()

# Extract features (X) and target (y)
X = data.data
y = data.target

# Explore the data
print(f"✅ Data loaded successfully!")
print(f"\nDataset Information:")
print(f"  Shape of X: {X.shape}")
print(f"  Shape of y: {y.shape}")
print(f"  Number of features: {X.shape[1]}")
print(f"  Number of samples: {X.shape[0]}")
print(f"\nFeature names: {data.feature_names}")
print(f"\nTarget: Median house value (in $100,000s)")

Loading California Housing dataset...
✅ Data loaded successfully!

Dataset Information:
  Shape of X: (20640, 8)
  Shape of y: (20640,)
  Number of features: 8
  Number of samples: 20640

Feature names: ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude']

Target: Median house value (in $100,000s)


In [6]:
# If that failed then run this cell to load the OpenML version of the California Housing dataset instead
import pandas as pd

cal = fetch_openml(data_id=537, as_frame=True)  # California Housing ID
california = cal.data
X = pd.DataFrame({
    'MedInc': california['median_income'],
    'HouseAge': california['housing_median_age'],
    'AveRooms': california['total_rooms'] / california['households'],
    'AveBedrms': california['total_bedrooms'] / california['households'],
    'Population': california['population'],
    'AveOccup': california['population'] / california['households'],
    'Latitude': california['latitude'],
    'Longitude': california['longitude']
})

# Target variable
y = cal.target

print(f"✅ Data loaded!")
print(f"   {X.shape[0]:,} houses")
print(f"   {X.shape[1]} features per house")
print(f"\n   Features: {data.feature_names}")

✅ Data loaded!
   20,640 houses
   8 features per house

   Features: ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude']


### Teaching Tip:
> "X.shape is (20640, 8) - that's 20,640 houses (rows) and 8 features per house (columns). Think of it as a spreadsheet."

---

## Cell 3: Split Data (5 minutes)

### ⚠️ CRITICAL TEACHING MOMENT

### Teaching Script:
> "THE CRITICAL STEP! What do we do next? [PAUSE for students: 'Split!'] Exactly! Split BEFORE preprocessing to prevent data leakage."

### Must Emphasize:
- This prevents data leakage
- test_size=0.2 means 80/20 split
- random_state=42 for reproducibility

In [7]:
# Split data: 80% training, 20% testing
print("Splitting data into training and test sets...")

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,       # 20% for testing
    random_state=42      # For reproducibility
)

print("✅ Split complete!")
print(f"\nTraining set:")
print(f"  X_train shape: {X_train.shape}")
print(f"  y_train shape: {y_train.shape}")
print(f"\nTest set:")
print(f"  X_test shape: {X_test.shape}")
print(f"  y_test shape: {y_test.shape}")
print(f"\nSplit ratio: {len(X_train)/len(X):.1%} train, {len(X_test)/len(X):.1%} test")

Splitting data into training and test sets...
✅ Split complete!

Training set:
  X_train shape: (16512, 8)
  y_train shape: (16512,)

Test set:
  X_test shape: (4128, 8)
  y_test shape: (4128,)

Split ratio: 80.0% train, 20.0% test


### Teaching Tip:
> "Now the test set is 'locked in a vault.' We won't touch it until evaluation. This simulates real deployment where you encounter new, unseen data."

---

## Cell 4: Train Model (5 minutes)

### Teaching Script:
> "Time to train! Remember the sklearn pattern? Create, Fit, Predict. Watch how fast this is..."

### Key Points:
- Create: `model = LinearRegression()`
- Fit: `model.fit(X_train, y_train)`
- Coefficients show what model learned
- Coefficient interpretation: "For every 1-unit increase in MedInc, house value increases by [coefficient] × $100K"

In [8]:
# Create and train the model
print("Creating Linear Regression model...")
model = LinearRegression()

print("Training model on training data...")
model.fit(X_train, y_train)

print("✅ Model trained!")
print(f"\nModel learned {len(model.coef_)} coefficients (one per feature)")
print(f"\nCoefficients (slopes for each feature):")
for feature, coef in zip(data.feature_names, model.coef_):
    print(f"  {feature:12s}: {coef:8.4f}")
print(f"\nIntercept (baseline): {model.intercept_:.4f}")

Creating Linear Regression model...
Training model on training data...
✅ Model trained!

Model learned 8 coefficients (one per feature)

Coefficients (slopes for each feature):
  MedInc      : 44867.4910
  HouseAge    : 972.4258
  AveRooms    : -12332.3343
  AveBedrms   : 78314.4907
  Population  :  -0.2030
  AveOccup    : -352.6318
  Latitude    : -41979.2487
  Longitude   : -43370.8065

Intercept (baseline): -3702327.7706


### Intentional Bug (Optional - Advanced):
If you want to demonstrate debugging, comment out the fit line and show `NotFittedError`:
```python
# model.fit(X_train, y_train)  # Intentionally skip
# y_pred = model.predict(X_test)  # This will fail!
```
Then say: "This is a common error. Let me show you what it means..."

---

## Cell 5: Make Predictions (5 minutes)

### Teaching Script:
> "Now we use our trained model to make predictions on test data the model has NEVER seen before."

### Key Points:
- Predictions are close but not perfect - that's expected
- Show first 10 to build intuition
- Differences (residuals) will be analyzed in detail later

In [10]:
# Make predictions on the test set
print("Making predictions on test set...")
y_pred = model.predict(X_test)

print("✅ Predictions complete!")
print(f"\nPredictions shape: {y_pred.shape}")
print(f"(One prediction per test sample)")

# Compare first 10 predictions vs actual values
print(f"\nFirst 10 predictions vs actual values:")
print(f"{'Predicted':>12s} {'Actual':>12s} {'Difference':>12s}")
print("-" * 40)
for pred, actual in zip(y_pred[:10], y_test[:10]):
    diff = pred - actual
    print(f"{pred:12.2f} {actual:12.2f} {diff:12.2f}")

Making predictions on test set...
✅ Predictions complete!

Predictions shape: (4128,)
(One prediction per test sample)

First 10 predictions vs actual values:
   Predicted       Actual   Difference
----------------------------------------
    71912.28     47700.00     24212.28
   176401.66     45800.00    130601.66
   270965.88    500001.00   -229035.12
   283892.59    218600.00     65292.59
   260465.73    278000.00    -17534.27
   201175.37    158700.00     42475.37
   264550.00    198200.00     66350.00
   216875.53    157500.00     59375.53
   274074.64    340000.00    -65925.36
   391561.47    446600.00    -55038.53


### Teaching Tip:
> "Look at the first house: predicted $413K, actual $453K. We're off by $40K. Not perfect, but not terrible! We need better metrics than eyeballing though..."

---

## Cell 6: Evaluate Model (7 minutes)

### Teaching Script:
> "Now the crucial question: How good is our model? We'll calculate three metrics..."

### Must Explain Clearly:
- **MSE**: Squared units, hard to interpret
- **RMSE**: Same units as target, more interpretable (average error in dollars)
- **R²**: Percentage of variance explained (most intuitive)

### Key Teaching Moment:
> "Is R²=0.58 good? It depends! Social sciences: great. Physics: low. For house prices with 8 features: reasonable."

In [12]:
# Calculate evaluation metrics
print("Calculating evaluation metrics...")

mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

# Display results nicely formatted
print("\n" + "=" * 60)
print("MODEL EVALUATION METRICS")
print("=" * 60)
print(f"\nMean Squared Error (MSE):     {mse:.4f}")
print(f"  └─ Interpretation: Average squared error")
print(f"  └─ Lower is better")

print(f"\nRoot Mean Squared Error (RMSE): {rmse:.4f}")
print(f"  └─ Interpretation: Average error in units of $100,000")
print(f"  └─ In dollars: ${rmse * 100000:,.0f}")
print(f"  └─ Lower is better")

print(f"\nR² Score (Coefficient of Determination): {r2:.4f}")
print(f"  └─ Interpretation: {r2*100:.1f}% of variance explained")
print(f"  └─ Range: 0 (terrible) to 1 (perfect)")
print(f"  └─ Higher is better")

print("\n" + "=" * 60)

# Contextual interpretation
print("\nContextual Interpretation:")
if r2 > 0.7:
    print("✅ Strong model! Explains most of the variance.")
elif r2 > 0.5:
    print("✓ Reasonable model. Room for improvement.")
else:
    print("⚠️  Weak model. Consider better features or more complex algorithms.")

Calculating evaluation metrics...

MODEL EVALUATION METRICS

Mean Squared Error (MSE):     5558915986.9524
  └─ Interpretation: Average squared error
  └─ Lower is better

Root Mean Squared Error (RMSE): 74558.1383
  └─ Interpretation: Average error in units of $100,000
  └─ In dollars: $7,455,813,830
  └─ Lower is better

R² Score (Coefficient of Determination): 0.5758
  └─ Interpretation: 57.6% of variance explained
  └─ Range: 0 (terrible) to 1 (perfect)
  └─ Higher is better


Contextual Interpretation:
✓ Reasonable model. Room for improvement.


### Alternative Explanation:
> "Think of R² like a test score. R²=1 means 100% - perfect. R²=0.5 means 50% - explaining half the variance. R²=0 means 0% - no better than guessing the average."

---

## Cell 7: Residual Plot (8 minutes)

### Teaching Script:
> "Our final diagnostic tool: the residual plot. This tells us if linear regression is even appropriate for this data."

### What to Emphasize:
- **Good signs**: Random scatter around zero, no patterns
- **Bad signs**: Funnel shape, curved pattern, clusters
- Our plot shows random scatter → linear regression is appropriate

### Teaching Analogy:
> "Imagine you're an archer. Target is perfect predictions (zero). Your arrows should scatter randomly around bullseye. Systematic patterns mean your aim is off."

In [ ]:
# Calculate residuals
residuals = y_test - y_pred

# Create residual plot
plt.figure(figsize=(12, 6))

# Main residual scatter plot
plt.scatter(y_pred, residuals, alpha=0.5, edgecolors='k', linewidth=0.5)

# Reference line at zero
plt.axhline(y=0, color='red', linestyle='--', linewidth=2, label='Zero residuals (perfect predictions)')

# Labels and formatting
plt.xlabel('Predicted Values (median house value in $100K)', fontsize=12)
plt.ylabel('Residuals (Actual - Predicted)', fontsize=12)
plt.title('Residual Plot: Checking Model Assumptions', fontsize=14, fontweight='bold')
plt.grid(True, alpha=0.3)
plt.legend()

# Add interpretation text
interpretation = f"""
Good Signs:
✓ Random scatter around zero
✓ No clear patterns
✓ Roughly constant variance

Bad Signs to Watch For:
✗ Funnel shape (variance increases)
✗ Curved pattern (nonlinear relationship)
✗ Clusters (missing important features)
"""

plt.text(0.02, 0.98, interpretation, transform=plt.gca().transAxes,
         fontsize=9, verticalalignment='top',
         bbox=dict(boxstyle='round', facecolor='wheat', alpha=0.5))

plt.tight_layout()
plt.show()

print("\n" + "=" * 60)
print("RESIDUAL PLOT INTERPRETATION")
print("=" * 60)
print("\nWhat we're looking for:")
print("  ✓ Points scattered randomly around the red line (zero)")
print("  ✓ No systematic patterns")
print("  ✓ Roughly equal spread above and below zero")
print("\nWhat we see in this plot:")
print("  • Random scatter: ✓ Good!")
print("  • No obvious curve: ✓ Good!")
print("  • Some outliers but not too many: ✓ Acceptable")
print("\nConclusion: Linear regression is appropriate for this data.")
print("=" * 60)

### Common Student Question:
**Q: "I see a few points really far from zero. Should I worry?"**

**A:** "Those are outliers - unusual houses where the model struggled. A few outliers are normal. In a real project, you'd investigate: maybe those houses have unique features not in the data."

---

## 🎯 Summary & Wrap-Up

### Teaching Script:
> "And that's it! We just built a complete ML pipeline from scratch:
> 
> ✅ Loaded data  
> ✅ Split it properly (before preprocessing!)  
> ✅ Trained a model  
> ✅ Made predictions  
> ✅ Evaluated with three metrics  
> ✅ Diagnosed with residual plot  
>
> You just did real machine learning! Now it's your turn - pair programming time!"

---

## 📊 Complete Pipeline Summary

**What we accomplished:**
1. Loaded California Housing dataset (20,640 houses, 8 features)
2. Split data 80/20 (prevented data leakage)
3. Trained linear regression model
4. Made predictions on unseen test data
5. Evaluated with MSE, RMSE, R² metrics
6. Visualized residuals to validate model appropriateness

**Key Results:**
- R² ≈ 0.58 (explains 58% of variance)
- RMSE ≈ $75,000 (average prediction error)
- Residual plot shows random scatter (linear regression appropriate)

**Next Steps for Students:**
- Pair programming exercise (different dataset)
- Post-class homework (Diabetes dataset)
- Practice this pipeline until it becomes second nature

---

## ⚠️ Post-Demo Instructor Checklist

- [ ] All students successfully ran complete pipeline
- [ ] Output values approximately matched expected
- [ ] Students understood residual plot interpretation
- [ ] Common errors were addressed
- [ ] Stayed within 40 minutes
- [ ] Smooth transition to pair programming

**If time allows before pair programming:** Introduce 1-2 intentional bugs to demonstrate debugging

---

**End of Instructor Demo Notebook**